grug/testing/src/client.rs

```diff
@@ -0,0 +1,313 @@
+        self.suite.lock().await.block_time = block_time;
+    }
+}
+
+#[async_trait]
+impl<DB, VM, PP, ID> QueryClient for MockClient<DB, VM, PP, ID>
+where
+    DB: Db + Send + Sync,
+    VM: Vm + Clone + Send + Sync + 'static,
+    PP: ProposalPreparer + Send + Sync,
+    ID: Indexer + Send + Sync,
+    AppError: From<DB::Error> + From<VM::Error> + From<PP::Error> + From<ID::Error>,
+{
+    type Error = anyhow::Error;
+    type Proof = grug_jmt::Proof;
+
+    async fn query_app(
+        &self,
+        query: Query,
+        height: Option<u64>,
+    ) -> Result<QueryResponse, Self::Error> {
+        Ok(self
+            .suite
+            .lock()
+            .await
+            .app
+            .do_query_app(query, height.unwrap_or_default(), false)?)
+    }
+
+    async fn query_store(
+        &self,
+        key: Binary,
+        height: Option<u64>,
+        prove: bool,
+    ) -> Result<(Option<Binary>, Option<Self::Proof>), Self::Error> {
+        let (value, proof) =
+            self.suite
+                .lock()
+                .await
+                .app
+                .do_query_store(&key, height.unwrap_or_default(), prove)?;
+
+        let value = value.map(Binary::from_inner);
+        let proof = proof.map(|p| p.deserialize_borsh()).transpose()?;
+
+        Ok((value, proof))
+    }
+
+    async fn simulate(&self, tx: UnsignedTx) -> Result<TxOutcome, Self::Error> {
+        Ok(self.suite.lock().await.app.do_simulate(tx, 0, false)?)
+    }
+}
+
+#[async_trait]
+impl<DB, VM, PP, ID> BlockClient for MockClient<DB, VM, PP, ID>
+where
+    DB: Db + Send + Sync,
+    VM: Vm + Send + Sync,
+    PP: ProposalPreparer + Send + Sync,
+    ID: Indexer + Send + Sync,
+{
+    type Error = anyhow::Error;
+
+    async fn query_block(&self, height: Option<u64>) -> Result<Block, Self::Error> {
+        let maybe_block = match height {
+            Some(height) => self
+                .blocks
+                .lock()
+                .await
+                .get(&height)
+                .map(|(block, _)| block.clone()),
+            None => self
+                .blocks
+                .lock()
+                .await
+                .last_key_value()
+                .map(|(_, (block, _))| block.clone()),
+        };
+
+        maybe_block.ok_or(anyhow::anyhow!("block not found: {:?}", height))
+    }
+
+    async fn query_block_outcome(&self, height: Option<u64>) -> Result<BlockOutcome, Self::Error> {
+        let maybe_block = match height {
+            Some(height) => self
+                .blocks
+                .lock()
+                .await
+                .get(&height)
+                .map(|(_, block)| block.clone()),
+            None => self
+                .blocks
+                .lock()
+                .await
+                .last_key_value()
+                .map(|(_, (_, block))| block.clone()),
+        };
+
+        maybe_block.ok_or(anyhow::anyhow!("block not found: {:?}", height))
```

Comment:
the message should be "block outcome not found" to differentiate with `query_block`